67_RayQueryGeometry/main.cpp

```diff
@@ -590,6 +597,235 @@ class RayQueryGeometryApp final : public examples::SimpleWindowedApplication, pu
+					IGPUBuffer::EUF_ACCELERATION_STRUCTURE_BUILD_INPUT_READ_ONLY_BIT | IGPUBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT;
+
+				auto iBuffer = smart_refctd_ptr(geom.data.indexBuffer.buffer); // no offset
+				auto iUsage = bitflag(IGPUBuffer::EUF_STORAGE_BUFFER_BIT) | IGPUBuffer::EUF_TRANSFER_DST_BIT | IGPUBuffer::EUF_INLINE_UPDATE_VIA_CMDBUF |
+					IGPUBuffer::EUF_ACCELERATION_STRUCTURE_BUILD_INPUT_READ_ONLY_BIT | IGPUBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT;
+
+				vBuffer->addUsageFlags(vUsage);
+				vBuffer->setContentHash(vBuffer->computeContentHash());
+				scratchObj.vertex = { .offset = 0, .buffer = vBuffer };
+
+				if (useIndex)
+					if (iBuffer)
+					{
+						iBuffer->addUsageFlags(iUsage);
+						iBuffer->setContentHash(iBuffer->computeContentHash());
+					}
+				scratchObj.index = { .offset = 0, .buffer = iBuffer };
+			}
+
+			// get ICPUBuffers into ICPUBottomLevelAccelerationStructures
+			std::array<smart_refctd_ptr<ICPUBottomLevelAccelerationStructure>, OT_COUNT> cpuBlas;
+			for (uint32_t i = 0; i < cpuBlas.size(); i++)
+			{
+				auto triangles = make_refctd_dynamic_array<smart_refctd_dynamic_array<ICPUBottomLevelAccelerationStructure::Triangles<ICPUBuffer>>>(1u);
+				auto primitiveCounts = make_refctd_dynamic_array<smart_refctd_dynamic_array<uint32_t>>(1u);
+
+				auto& tri = triangles->front();
+				auto& primCount = primitiveCounts->front();
+				const auto& geom = objectsCpu[i];
+				const auto& scratchObj = scratchBuffers[i];
+
+				const bool useIndex = geom.data.indexType != EIT_UNKNOWN;
+				const uint32_t vertexStride = geom.data.inputParams.bindings[0].stride;
+				const uint32_t numVertices = scratchObj.vertex.buffer->getSize() / vertexStride;
+
+				if (useIndex)
+					primCount = geom.data.indexCount / 3;
+				else
+					primCount = numVertices / 3;
+
+				geomInfos[i].indexType = geom.data.indexType;
+				geomInfos[i].vertexStride = vertexStride;
+				geomInfos[i].smoothNormals = smoothNormals[i];
+
+				tri.vertexData[0] = scratchObj.vertex;
+				tri.indexData = useIndex ? scratchObj.index : scratchObj.vertex;
+				tri.maxVertex = numVertices - 1;
+				tri.vertexStride = vertexStride;
+				tri.vertexFormat = EF_R32G32B32_SFLOAT;
+				tri.indexType = geom.data.indexType;
+				tri.geometryFlags = IGPUBottomLevelAccelerationStructure::GEOMETRY_FLAGS::OPAQUE_BIT;
+
+				auto& blas = cpuBlas[i];
+				blas->setGeometries(std::move(triangles), std::move(primitiveCounts));
+
+				auto blasFlags = bitflag(IGPUBottomLevelAccelerationStructure::BUILD_FLAGS::PREFER_FAST_TRACE_BIT) | IGPUBottomLevelAccelerationStructure::BUILD_FLAGS::ALLOW_COMPACTION_BIT;
+				if (m_physicalDevice->getProperties().limits.rayTracingPositionFetch)
+					blasFlags |= IGPUBottomLevelAccelerationStructure::BUILD_FLAGS::ALLOW_DATA_ACCESS;
+
+				blas->setBuildFlags(blasFlags);
+				blas->setContentHash(blas->computeContentHash());
+			}
+
+			// TODO: when does compact blas happen?
```

Comment:
during conversion, you just need to make sure that you provide a `input.patches` with `compactAfterBuild` instead of leaving an empty default span

Reply:
set it only for the TLAS